app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/ui/activities/ProfileEditingActivity.java

```diff
@@ -48,28 +55,45 @@ protected void onCreate(Bundle savedInstanceState) {
         View view = binding.getRoot();
         setContentView(view);
         vmodel = new ViewModelProvider(this, new EditProfileViewModelFactory((NeighborFoodApplication) this.getApplication())).get(EditProfileViewModel.class);
-        vmodel.getCurrentUser().observe(this,user -> {
-            updateUserFields(user);
-        });
+        vmodel.getCurrentUser().observe(this, this::updateUserFields);
         setContentView(R.layout.activity_profile_editing);
         Toolbar toolbar = findViewById(R.id.profileEditToolbar);
         setSupportActionBar(toolbar);
-
         ppView = findViewById(R.id.profilePictureImageView);
         ppView.setOnClickListener(this::onClick);
-
+        findViewById(R.id.profileEditAddLinkButton).setOnClickListener(this::onClick);
+        linksLayout = (LinearLayout) findViewById(R.id.profileEditLinksLayout);
+        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
     }
 
-    public void updateUserFields(User user) {
-        if(user != null){
-            ((TextView)findViewById(R.id.nameValue)).setText(user.getFirstName());
-            ((TextView)findViewById(R.id.surnameValue)).setText(user.getLastName());
-            ((TextView)findViewById(R.id.emailValue)).setText(user.getEmail());
-            Picasso.with(this).load(user.getProfilePictureURI()).fit().into(ppView);
+    private void updateUserFields(User user) {
+        if(user == null){
+            return;
         }
+        ((TextView)findViewById(R.id.nameValue)).setText(user.getFirstName());
+        ((TextView)findViewById(R.id.surnameValue)).setText(user.getLastName());
+        ((TextView)findViewById(R.id.emailValue)).setText(user.getEmail());
+        ((TextView)findViewById(R.id.bioValue)).setText(user.getBio());
+        Picasso.with(this).load(user.getProfilePictureURI()).fit().into(ppView);
+        updateUserLinks(user);
 
     }
 
+    private void updateUserLinks(User user){
+        //TODO: add checks to not reinstantite every time, keep track for more optimization
+
+        //discard all previous TextInputEditText
+        linksLayout.removeAllViews();
+        //add the links the user already has
+        textEdits = new ArrayList<>();
+        for(String s : user.getLinks() ){
+            TextInputEditText t = addLinkInput();
+            t.setText(s);
+        }
```

Comment:
maybe there's a better way for updating the links instead of removing them all then adding them one by one but this is the easiest way I can see for now since link fields can be updated randomly and have new ones.
(A possible solution is to add a state modified whenever a link is modified and only change modified links later in the save but that will be as costly as your method and harder to implement but a bit cleaner).

Reply:
I agree with you, I have even put a TODO comment to change how the views are put, if it makes us gain more performance, but I think It doesn't really matter for now, as I haven't seen any performance issues especially for this feature, and i am afraid that  implementing it the other way might be much of a headache for nothing